src/authorization/permissions/IntermediateModesExtractor.ts

```diff
@@ -0,0 +1,54 @@
+import type { Operation } from '../../http/Operation';
+import type { ResourceSet } from '../../storage/ResourceSet';
+import type { IdentifierStrategy } from '../../util/identifiers/IdentifierStrategy';
+import { ModesExtractor } from './ModesExtractor';
+import type { AccessMap } from './Permissions';
+import { AccessMode } from './Permissions';
+import { addAccessModes } from './PermissionUtil';
+
+/**
+ * Returns the required access modes from the source {@link ModesExtractor}.
+ * In case create permissions are required,
+ * verifies if any of the containers permissions also need to be created
+ * and adds the corresponding identifier/mode combinations.
+ */
+export class IntermediateModesExtractor extends ModesExtractor {
+  private readonly resourceSet: ResourceSet;
+  private readonly strategy: IdentifierStrategy;
+  private readonly source: ModesExtractor;
+
+  /**
+   * Certain permissions depend on the existence of the target resource.
+   * The provided {@link ResourceSet} will be used for that.
+   * @param resourceSet - {@link ResourceSet} that can verify the target resource existence.
+   * @param strategy - {@link IdentifierStrategy} that will be used to determine parent containers.
+   * @param source - The source {@link ModesExtractor}.
+   */
+  public constructor(resourceSet: ResourceSet, strategy: IdentifierStrategy, source: ModesExtractor) {
+    super();
+    this.resourceSet = resourceSet;
+    this.strategy = strategy;
+    this.source = source;
+  }
+
+  public async canHandle(input: Operation): Promise<void> {
+    return this.source.canHandle(input);
+  }
+
+  public async handle(input: Operation): Promise<AccessMap> {
+    const accessMap = await this.source.handle(input);
+
+    for (const [ key, modes ] of accessMap) {
+      if (modes.has(AccessMode.create)) {
+        // Add create AccessMode if parent does not exist yet
+        const parent = this.strategy.getParentContainer(key);
+        if (!await this.resourceSet.hasResource(parent)) {
+          // Don't need to add write/append permissions as those are already being checked on the final resource
```

Comment:
by whom?

Reply:
Well that depends on the configured permission readers 😛.

Actually spend some time trying to answer the question which permissions should be added here. E.g., does write also need to be added here? Does append? The interesting thing is that we currently require different permissions when creating a resource using PUT or PATCH (can definitely discuss if this is correct, but it is one way to interpret the current solid spec), so would we need to add the same one here or not?

In hindsight the comment/solution here is not completely correct since it is sort of based on how ACL authorization works. In practice the same permissions will be applied to the intermediate containers as on the final target identifier, since all of these are non-existent resources that will all use the permissions of the parent container in which they are created. If the user has no write permissions on that container, they will not be allowed to create the final resource so the request will be denied and the intermediate containers will also not be created. So it doesn't matter if we also request write/append permissions for them. But that idea is perhaps a bit too ACL-minded.

Question then is which modes to request here. Could of course just stick with the create but change the comment.

Reply:
Comment sounds good to me.